File: 200.number-of-islands.go
package main

import "fmt"

func main() {
	fmt.Println(numIslands([][]byte{[]byte{'1', '0', '1', '1', '0', '1', '1'}}))
}

func numIslands(grid [][]byte) int {
	if len(grid) == 0 || len(grid[0]) == 0 {
		return 0
	}
	visited := make([][]bool, len(grid))
	for i := range visited {
		visited[i] = make([]bool, len(grid[0]))
	}
	mxcnt := 0
	for i := range grid {
		for j := range grid[i] {
			if visited[i][j] {
				continue
			}
			if grid[i][j] == '1' {
				mxcnt++
				spread(i, j, grid, visited)
			}
		}
	}
	return mxcnt
}

func spread(i, j int, grid [][]byte, visited [][]bool) {
	if i < 0 || j < 0 || i >= len(grid) || j >= len(grid[0]) || visited[i][j] || grid[i][j] == '0' {
		return
	}
	if grid[i][j] == '1' {
		visited[i][j] = true
	}
	spread(i-1, j, grid, visited)
	spread(i, j-1, grid, visited)
	spread(i+1, j, grid, visited)
	spread(i, j+1, grid, visited)
}

File: 200.number-of-islands.java
class Solution {

    static class Pair {
        public int l;
        public int r;

        public Pair(int l, int r) {
            this.l = l;
            this.r = r;
        }
    }

    public int numIslands(char[][] grid) {
        Queue<Pair> queue = new LinkedList<>();
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int ans = 0;
        int[] dx = new int[]{0, 1, 0, -1};
        int[] dy = new int[]{-1, 0, 1, 0};
        
        for (int i = 0; i < visited.length; i++) {
            for (int j = 0; j < visited[0].length; j++) {
                if (visited[i][j]) {
                    continue;
                }
                if (grid[i][j] != '1') {
                    visited[i][j] = true;
                    continue;
                }
                ans ++;
                queue.add(new Pair(i, j));
                while (!queue.isEmpty()) {
                    Pair p = queue.poll();
                    if (visited[p.l][p.r]) {
                        continue;
                    }
                    if (grid[p.l][p.r] != '1') {
                        continue;
                    }
                    visited[p.l][p.r] = true;
                    for (int di = 0; di < 4; di++) {
                        int nl = p.l + dx[di];
                        int nr = p.r + dy[di];

                        if (nl >= 0 && nl < grid.length && nr >= 0 && nr < grid[0].length) {
                            queue.add(new Pair(nl, nr));
                        }
                    }
                }
            }
        }
        return ans;

    }
}

